vql/linux/audit.go

```diff
@@ -4,90 +4,688 @@ package linux
+	watcher.disconnect()
+
+	self.watcherLock.Lock()
+
+	_, ok := self.watchers[watcher.id]
+	if ok {
+		delete(self.watchers, watcher.id)
+
+		self.watcherLock.Unlock()
+		self.wgDec("connectWatcher")
+		self.Put()
+	} else {
+		self.watcherLock.Unlock()
+	}
+}
+
+func (self *AuditWatcherService) disconnectWatchers(watchers []*AuditWatcher) {
+	defer self.wgDec("disconnectWatchers")
+
+	for _, watcher := range watchers {
+		self.disconnectWatcher(watcher)
+	}
+}
+
+func (self *AuditWatcherService) distributeEvent(event *aucoalesce.Event) {
+
+	watchersToDisconnect := []*AuditWatcher{}
+
+	self.watcherLock.RLock()
+	for _, watcher := range self.watchers {
+		select {
+		case <- watcher.ctx.Done():
+			watchersToDisconnect = append(watchersToDisconnect, watcher)
+		case watcher.eventChannel <- *event:
+		}
+	}
+	self.watcherLock.RUnlock()
+
+	if len(watchersToDisconnect) > 0 {
+		self.wgInc("disconnectWatchers")
+		go self.disconnectWatchers(watchersToDisconnect)
+	}
+}
+
+func (self *AuditWatcherService) notifyWatchers(message string) {
+	self.watcherLock.RLock()
+	defer self.watcherLock.RUnlock()
+
+	for _, watcher := range self.watchers {
+		watcher.scope.Log(message)
+	}
+}
+
+func (self *AuditWatcherService) notifyMissingRule(rule *AuditWatcherRule) {
+	self.watcherLock.RLock()
+	defer self.watcherLock.RLock()
+	count := 0
+
+	for _, watcher := range self.watchers {
+		_, ok := watcher.rules[rule.rule]
+		if ok {
+			watcher.scope.Log("audit: replaced missing rule `%v'", rule.rule)
+			count += 1
+		}
+	}
+
+	if count > 0 {
+		self.logger.Info("audit: replaced missing rule `%v'", rule.rule)
 	}
 }
 
-func addRules(c *libaudit.AuditClient, rulelst []string, s vfilter.Scope) (error) {
-       for _, line := range rulelst {
-              r, err := flags.Parse(line)
-              if err != nil {
-                     s.Log("Error parsing Rule %v: %s", err)
-                     return fmt.Errorf("Error while adding rules: %w", err)
-               }
-               data, err := rule.Build(r)
-               if err != nil {
-                      s.Log("Error Building Rule: %s", err)
-                      return fmt.Errorf("Error while adding rules: %w", err)
-               }
-               defer c.GetRules()
-               //defer deleteRules(c)
-               if err := c.AddRule([]byte(data)); err != nil {
-                       if strings.Contains(err.Error(), "rule exists"){
-                              continue
-                       } else {
-                              s.Log("Error while adding Rule %s: %s", line,err)
-                              return fmt.Errorf("error adding audit rule: %w", err)
-                       }
+func (self *AuditWatcherService) checkRules() error {
+	self.rulesLock.Lock()
+	defer self.rulesLock.Unlock()
+
+	self.notifyWatchers("audit: detected audit configuration change")
+	self.logger.Info("audit: detected audit configuration change")
```

Comment:
The call to self.logger.Info is redundant since notifyWatchers will cause every watched to print the message. This results in duplicated message printout: 

```
[INFO] 2022-06-10T10:54:41Z Compiled all artifacts. 
[INFO] 2022-06-10T10:54:41Z audit: detected audit configuration change 
[INFO] 2022-06-10T10:54:41Z audit: detected audit configuration change 
```

Reply:
Yep. It'll always result in duplicate message printout. In fact, there'll be a message for every watcher. IMO this is a problem with Velociraptor. If we don't log it multiple times, it won't appear in the log in the UI for the plugin. I get that you're saying that it is separate from the scope, but I don't think one more entry hurts when there will already be several. Eventually, I'd like to fix it so that the logging works differently.